src/helpers/authed-server-side-props.ts

```diff
@@ -0,0 +1,51 @@
+import { GetServerSideProps, GetServerSidePropsContext } from 'next';
+import { User } from 'src/domain/user';
+import {
+  RequestAuthorizer,
+  RequestAuthorizerCommand,
+} from 'src/services/request-authorizer';
+import { AuthenticationError } from '../../types/auth-errors';
+
+const authorizer = new RequestAuthorizer();
+
+type AuthProps = { user?: User };
+export type WithUser<PropsType = void> = PropsType & AuthProps;
+
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export function withAuth<PropsType extends { [key: string]: unknown }>(
+  gssp?: GetServerSideProps<PropsType>
+) {
+  const authedGssp = async (ctx: GetServerSidePropsContext) => {
+    const command: RequestAuthorizerCommand = {
+      path: ctx.resolvedUrl,
+      cookieHeader: ctx.req?.headers.cookie,
+    };
+
+    const authResult = authorizer.execute(command);
+
+    if (authResult.success) {
+      if (gssp === undefined) {
+        return {
+          props: ({ user: authResult.user } as unknown) as WithUser<PropsType>,
+        };
+      }
+
+      const componentProps = await gssp(ctx);
+
+      if (!('props' in componentProps)) return componentProps;
```

Comment:
Could you explain what this line does?

Reply:
So this is checking whether the `getServerSideProps` function (defined from the component this function is wrapping) returned an object with `props` on it. Recalling the [`getServerSideProps` return format](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering), it may return props (i.e. a successful function call) or it may return `redirect` or a number of other things.

If it doesn't return `props`, we want to just exit out early and return whatever it returned so that Next.js can do its thing. If it does return `props`, we want to add in the `user` before returning.

The reason it's `'props' in componentProps` and not `componentProps.props` that Next.js defines the return type of `getServerSideProps` as 
```
export type GetServerSidePropsResult<P> =
  | { props: P }
  | { redirect: Redirect }
  | { notFound: true }
```
and because `props` doesn't exist on all of the possible return types, typescript doesn't let us access it normally. the `[]` accessor effectively bypasses the type checking on that.

Reply:
Makes sense, thanks 👍🏻